### ISSUE TITLE
Batched sparse-sparse matrix multiplication/ sparse torch.einsum

### ISSUE THREAD

**Issue by lpxhonneux** [pytorch/pytorch#72065]
### 🚀 The feature, motivation and pitch

I want to do a sort of batched sparse-sparse matrix multiplication, but specifically in the scenario where I have two hybrid sparse coo tensors of the shape (N,N,C), where only the C dimension is a dense tensor. The behaviour I am looking for is that of a batched sparse-sparse matrix multiplication where the C dimension is in effect the batch dimension. Or equivalently expressed in torch.einsum: "i k d, j k d -> i j d" but only over the sparse indices in the first two dimensions.

If there exists a torch.sparse.bmm() function, it would be possible to permute the indices and do this I think.

### Alternatives

The only alternative I can come up with is to allow torch.einsum to work on sparse operands and only consider operating the einsum over the existing sparse indices in the tensor.

### Additional context

I need the feature to implement the TriangleMultiplication (Line 4 of Alg. 11 in the Supplementary Material) from Alphafold2 but for graphs batched as one large disconnected graph as is common in pytorch_geometric.

**Comment by cpuhrsch** [comment]
Hey @lpxhonneux, we have support for torch.bmm(sparse, dense). It just behaves like the dense counterpart, i.e. commutes with to_dense. If we had support for torch.bmm(sparse, sparse) with the same semantics would that resolve your issue?

Also, which levels of sparsity do you expect? For example, will your SparseTensor be 99% sparse?

**Comment by lpxhonneux** [comment]
Hey, support for torch.bmm(sparse, sparse) should be sufficient functionally, but I think it might miss a lot of opportunity for vectorisation as the sparse matrix always has the same indices (i,j) but with different entries (all entries captured as a vector in the final dimension), i.e. it can be viewed as a single matrix multiplication with the entries of the matrix not being scalars but vectors that simply need to be multiplied element-wise. 

My SparseTensor is create by batching a set of graphs into a large disconnected graph represented as a SparseTensor (this is the standard approach in pytorch geometric, so a common use case of SparseTensors in pytorch I suspect). This means the sparse tensor is block diagonal, where each block on the diagonal may be sparse or dense depending on the type of graphs in the dataset. So assuming the graphs are roughly equal size and fully connected (the most extreme case), in which case the number of nonzero entries is 1/B (if my napkin math is correct), where B is the batchsize with a typical batchsize between 32-128, so between 3-0.5%.

Thank you for the quick response! Let me know if you need anything else.

**Comment by cpuhrsch** [comment]
@lpxhonneux - Ok that makes sense! Thanks for the quick reply!

I'd start out with torch.bmm(sparse, sparse) first then, since it's a very generic operation, and add the other operations you suggested as a performance optimization later on.

That way at least functionally your are unblock and we can use a benchmark snippet that's based on a concrete model within your work.

On that note, can you share a small snippet that we could run to check the performance of our kernels in the context of your work?

**Comment by lpxhonneux** [comment]
Yes, I will write such a snippet of code. I am quite busy today, but I'll try to get it done either this evening or tomorrow.

**Comment by cpuhrsch** [comment]
@lpxhonneux for what it's worth we already have support for mm(CSR, CSR) -> CSR

```
>>> a = torch.randn(3, 3).maximum(torch.tensor(0))
>>> b = torch.randn(3, 3).maximum(torch.tensor(0))
>>> a
tensor([[0.8409, 0.0000, 1.2554],
        [0.0000, 0.0531, 0.2517],
        [1.7138, 1.0580, 1.6732]])
>>> b
tensor([[0.0000, 0.5418, 0.0000],
        [0.0000, 0.0000, 0.0000],
        [1.2737, 0.6904, 0.0898]])
>>> torch.mm(a.to_sparse_csr(), b.to_sparse_csr()).to_dense()
tensor([[1.5990, 1.3223, 0.1127],
        [0.3206, 0.1738, 0.0226],
        [2.1311, 2.0836, 0.1502]])
>>> torch.mm(a, b)
tensor([[1.5990, 1.3223, 0.1127],
        [0.3206, 0.1738, 0.0226],
        [2.1311, 2.0836, 0.1502]])
```

We're still working out the details of batch support for CSR Tensors.

Are you open to using CSR (cuSPARSE has efficient CSR routines) or are you tied to COO?

**Comment by lpxhonneux** [comment]
Here is some code (tested on cpu in pytorch 1.8.1) that shows the style of input data and my current solutions:
```
import torch

def fully_connected_index(batch, num_nodes):
    node_range = torch.arange(0,num_nodes.sum())
    src = torch.repeat_interleave(node_range, torch.repeat_interleave(num_nodes,num_nodes))
    end = torch.cat([node_range[batch==i].repeat(n) for i,n in enumerate(num_nodes) ])
    return torch.stack([src,end], dim=0)

channel_dim = 128 # typical 16-128
batch_size = 32 # typical 16-128
num_nodes = torch.randint(50,200,(batch_size,))
edges = torch.randn(((num_nodes**2).sum(),channel_dim))
indices = fully_connected_index(
    torch.repeat_interleave(torch.arange(0,batch_size),num_nodes),
    num_nodes
)
print(edges.shape)

# I get nodes and index from previous code and can then create the sparse hybrid coo tensor I need
hybrid_tensor = torch.sparse_coo_tensor(indices,edges, size=(num_nodes.sum(),num_nodes.sum(),channel_dim))

# if torch.bmm() works on sparse x sparse,
# though how to deal with batch dimension being at the end is also non-trivial,
# ideally torch.bmm for sparse sparse could be batched in the last dimension to take advantage
# of the existing hybrid coo tensors
# ideally something like the line below would be possible, where the dim argument indices the batched dimension
# desired_output = torch.bmm(hybrid_tensor,hybrid_tensor.t(),dim=2).values().view(-1,channel_dim)

# Two ways I currently execute this, which works but require for loops/list comprehensions
hybrid_tensor = torch.sparse_coo_tensor(indices, edges, size=(num_nodes.sum(),num_nodes.sum(),channel_dim))
# Option A General and too slow to be viable
out_A = []
for i in range(channel_dim):
    tensor_i = torch.sparse_coo_tensor(indices, edges[:,i], size=(num_nodes.sum(),num_nodes.sum()))
    out_A.append(torch.sparse.mm(tensor_i, tensor_i.t()).coalesce().values())
out_A = torch.stack(out_A,dim=-1)
print(out_A.shape)

# Option B only when the graphs are fully connected as will usually be the case in my situation and much faster than A
out_B = []
cum_n_atoms_squared = 0
for i, n in enumerate(num_nodes):
    graph_i = edges[cum_n_atoms_squared:(cum_n_atoms_squared+n*n)].view(n,n,channel_dim)
    cum_n_atoms_squared += n*n
    out_B.append(torch.einsum("... i k d, ... j k d -> ... i j d", graph_i,graph_i).view(-1,channel_dim))
out_B = torch.cat(out_B,dim=0)
print(out_B.shape)
```

**Comment by ChungeonKim** [comment]
Did you solve this problem?

I have the same issue. I want to conduct bmm(sparse_matrix, sparse_vec) where sparse_matrix has the shape of (batch, N, N) and sparse_vec has the shape of (batch, N) or (batch, N, 1).

**Comment by bjornvz** [comment]
I have a similar issue: 

I use **torch.matmul( x, w )** with
- **x: (B, N)**, strided
- **w: (N, N)**, sparse

and **M*torch.matmul( x.T, y )** (element-wise product times batched outer product)
- **M: (N, N)** sparse
- **x: (B, N)**, strided
- **y: (B, N)**, strided

**Comment by RongqinChen** [comment]
Has this feature been implemented?

**Comment by thomasahle** [comment]
I have this problem too, and my sparse matrix is actually a sparse tensor.
Having einsum work would be really helpful, since bmm requires a matrix shape, and sparse tensors can't be reshaped :(
Right now I have to create 8 different versions of my sparse tensor, depending on which entries I end up needing to contract over.
